src/cell_list.h

```diff
@@ -0,0 +1,226 @@
+                         int distance ) : list( &list ), origin( p ), origin_cell( p / CELL_SIZE ),
+            max_distance( distance ), sorted_elements( distance + 1, std::vector<std::pair<point, T >>() ) {
+            min_bounds = origin_cell * CELL_SIZE;
+            max_bounds = min_bounds + point( CELL_SIZE - 1, CELL_SIZE - 1 );
+            distance_to_nearest_cell_wall = calc_distance();
+            // Special case for first cell.
+            add( list.cells[origin_cell.x][origin_cell.y] );
+        }
+        using iterator = cell_list_iterator<T, CELL_SIZE, MAXX, MAXY>;
+        iterator begin() {
+            return iterator( this );
+        }
+        iterator end() {
+            return iterator();
+        }
+        point get_origin() const {
+            return origin;
+        }
+        int get_distance() const {
+            return max_distance;
+        }
+        friend class cell_list_iterator< T, CELL_SIZE, MAXX, MAXY >;
+    private:
+        int calc_distance() const {
+            int distance = max_distance;
+            if( min_bounds.x >= 0 ) {
+                distance = std::min( distance, origin.x - min_bounds.x );
+            }
+            if( min_bounds.y >= 0 ) {
+                distance = std::min( distance, origin.y - min_bounds.y );
+            }
+            if( max_bounds.x < MAXX ) {
+                distance = std::min( distance, max_bounds.x - origin.x );
+            }
+            if( max_bounds.y < MAXY ) {
+                distance = std::min( distance, max_bounds.y - origin.y );
+            }
+            return distance;
+        }
+        void add( const std::vector<std::pair<point, T>> &cells ) {
+            for( const std::pair<point, T> &candidate : cells ) {
+                add( candidate );
+            }
+        }
+        void add( const std::pair<point, T> &candidate ) {
+            const int distance = rl_dist( candidate.first, origin );
+            if( distance > max_distance ) {
+                return;
+            }
+            sorted_elements[distance].push_back( candidate );
+        }
+        void load() {
+            constexpr point rank_offset{ CELL_SIZE, CELL_SIZE };
+            min_bounds -= rank_offset;
+            max_bounds += rank_offset;
+            distance_to_nearest_cell_wall = calc_distance();
+            cell_ranks_loaded++;
+            const point load_offset( cell_ranks_loaded, cell_ranks_loaded );
+            const point min_cell = origin_cell - load_offset;
+            const point max_cell = origin_cell + load_offset;
+            const bool min_x_inbounds = min_cell.x >= 0;
+            const bool min_y_inbounds = min_cell.y >= 0;
+            const bool max_x_inbounds = max_cell.x < list->num_cells_x();
+            const bool max_y_inbounds = max_cell.y < list->num_cells_x();
+            // Draw a box...
+            for( int x = min_cell.x; x <= max_cell.x; ++x ) {
+                if( x < 0 || x >= list->num_cells_x() ) {
```

Comment:
These checks can be brought out of the loop and embodied into `min_cell` / `max_cell` (same for `y`).

Reply:
I thought so at first, but I can't move either min_cell or max_cell because that would shift the box and make the behavior incorrect, e.g. if you have values like this
```
(-1, -1), (0, -1), ( 1, -1),
(-1,  0),              ( 1,  0),
(-1,  1), ( 0,  1), ( 1,  1)
```
Then adjusting min_cell and max_cell would result in:
```
( 0,  0), ( 1, 0),
( 0,  1), ( 1, 1)
```
Which is incorrect, what we want is
```
           (1, 0),
(0, 1), (1, 1)
```